src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx

```diff
@@ -0,0 +1,215 @@
+    IVisProps,
+    IBucketItem,
+    IBucket,
+} from "../../../interfaces/Visualization";
+import { PluggableBaseChart } from "../baseChart/PluggableBaseChart";
+import { BUCKETS, METRIC, ATTRIBUTE } from "../../../constants/bucket";
+import { GEO_PUSHPIN_CHART_UICONFIG } from "../../../constants/uiConfig";
+import {
+    sanitizeFilters,
+    getBucketItemsByType,
+    getPreferredBucketItems,
+    hasBucket,
+} from "../../../utils/bucketHelper";
+import { setGeoPushpinUiConfig } from "../../../utils/uiConfigHelpers/geoPushpinChartUiConfigHelper";
+import { removeSort, createSorts } from "../../../utils/sort";
+import { getReferencePointWithSupportedProperties } from "../../../utils/propertiesHelper";
+import { VisualizationTypes } from "../../../../constants/visualizationTypes";
+
+import UnsupportedConfigurationPanel from "../../configurationPanels/UnsupportedConfigurationPanel";
+import { DASHBOARDS_ENVIRONMENT } from "../../../constants/properties";
+import { GeoChart } from "../../../../components/core/GeoChart";
+
+export class PluggableGeoPushpinChart extends PluggableBaseChart {
+    private geoPushpinElement: string;
+
+    constructor(props: IVisConstruct) {
+        super(props);
+
+        const { callbacks, element, visualizationProperties } = props;
+        this.type = VisualizationTypes.PUSHPIN;
+        this.supportedPropertiesList = [];
+        this.callbacks = callbacks;
+        this.geoPushpinElement = element;
+        this.initializeProperties(visualizationProperties);
+    }
+
+    public getExtendedReferencePoint(referencePoint: IReferencePoint): Promise<IExtendedReferencePoint> {
+        const clonedReferencePoint = cloneDeep(referencePoint);
+        let newReferencePoint: IExtendedReferencePoint = {
+            ...clonedReferencePoint,
+            uiConfig: cloneDeep(GEO_PUSHPIN_CHART_UICONFIG),
+        };
+
+        const buckets = get(clonedReferencePoint, BUCKETS, []);
+        const locations = getBucketItemsByType(buckets, BucketNames.LOCATION, [ATTRIBUTE]);
+        const measuresSize = this.getMeasureSizeBucketItem(buckets);
+        const measuresColor = this.getMeasureColorBucketItem(buckets);
+        const segments = getPreferredBucketItems(
+            buckets,
+            [BucketNames.STACK, BucketNames.SEGMENT_BY],
+            [ATTRIBUTE],
+        );
+
+        set(newReferencePoint, BUCKETS, [
+            {
+                localIdentifier: BucketNames.LOCATION,
+                items: locations,
+            },
+            {
+                localIdentifier: BucketNames.SIZE,
+                items: measuresSize,
+            },
+            {
+                localIdentifier: BucketNames.COLOR,
+                items: measuresColor,
+            },
+            {
+                localIdentifier: BucketNames.SEGMENT_BY,
+                items: segments,
+            },
+        ]);
+
+        newReferencePoint = setGeoPushpinUiConfig(newReferencePoint, this.intl, this.type);
```

Comment:
This line and line [54-57](https://github.com/gooddata/gooddata-react-components/pull/1309/files#diff-75f470fd75494345ded6b8f852fcb9c0R54-R57) should be combined into one function. And place on top of this function.

Reply:
it required buckets information to update uiConfig. So keep it as is.